### ISSUE TITLE
[12] Split generator into focused modules after template extraction

### ISSUE THREAD

**Issue by wildersa** [wildersa/ml-starter-lab-kit#20]
# [12] Split generator into focused modules after template extraction

## Template evaluation metadata

This issue uses the template declared in the `jules_eval` block above.

The issue is the source of truth for template evaluation. The PR does not need to repeat `template_id`; it only needs to link this issue using GitHub closing keywords such as `Fixes #20` or `Closes #20`.

Evaluation will be based on the linked PR outcome, review comments, required rework, CI status, merge status, and whether follow-up issues were needed.

---

## Agent behavior instructions

You are a senior Python architecture and Machine Learning tooling engineer.

You specialize in:

- small Python CLI tools built only with the standard library;
- ML project scaffolding for practical projects using `pandas`, `numpy`, `scikit-learn`, `xgboost`, `matplotlib`, notebooks, baseline models, metrics, and optional PyTorch workflows;
- maintainable `src/`-layout generated projects with `pyproject.toml`, requirements files, tests, notebooks, configs, reports, and data folders;
- separating responsibilities in Python code without overengineering;
- preserving user-facing behavior during internal refactors;
- writing beginner-readable code that a student can inspect and modify.

Your job in this issue is to make the generator easier to maintain by separating responsibilities. This is an internal organization refactor, not a product redesign.

Before changing code:

- inspect the current generator structure after #19 is merged;
- inspect existing tests and helper structure after #18 is merged;
- inspect `templates/**` after template extraction;
- map current responsibilities before moving code;
- keep this PR focused on splitting responsibilities, not changing generated output.

Important attention points:

- Keep `python create_ml_starter.py` working as the main entrypoint.
- Keep the generator standard-library only. Do not add Click, Typer, Jinja, Pydantic, Cookiecutter, or any external dependency.
- Do not create a plugin framework, task registry, preset registry, dependency resolver, or architecture layer.
- Prefer a few clear modules over many tiny files.
- Preserve generated config schema, generated directory layout, generated import paths, and test behavior.
- Do not reintroduce Datathon behavior.
- Do not use this refactor to change ML task semantics or add new ML features.

This issue must run after:

- #18 — test helper simplification;
- #19 — generated file template extraction.

If those issues are not merged yet, do not implement this issue.

A PR that changes generated project behavior without explicit tests is incomplete.

---

## Problem

`create_ml_starter.py` is doing too many things in one file: CLI prompts, option selection, config building, directory creation, template rendering, generated file orchestration, optional file handling, and summary output.

Observed behavior:

```txt
The generator works, but the implementation is monolithic and becoming hard to maintain.
```

Expected behavior:

```txt
Generator responsibilities should be split into small focused modules/functions while preserving the existing CLI and generated output behavior.
```

The problem is believed to happen in:

- `create_ml_starter.py`

Do not solve this by redesigning the generator architecture broadly or introducing a framework.

---

## P0 objectives

The PR must satisfy all P0 items.

- [ ] P0.1 — Split generator responsibilities into focused modules or clearly separated internal components.
- [ ] P0.2 — Keep `create_ml_starter.py` as the CLI entrypoint so existing usage still works.
- [ ] P0.3 — Preserve existing supported generated project behavior.
- [ ] P0.4 — Preserve existing tests and add/update focused tests only where module extraction changes import paths or public helpers.
- [ ] P0.5 — Keep the generator standard-library only and beginner-readable.

A PR that does not satisfy every P0 item is incomplete, even if CI passes.

---

## Non-solutions

The following do not satisfy this issue:

- Rewriting the generator from scratch.
- Adding Click, Typer, Jinja, Pydantic, Cookiecutter, or any external dependency.
- Changing CLI behavior, prompt order, or generated output without explicit P0 need.
- Adding new project types, presets, environment profiles, or features.
- Mixing this with language cleanup, Datathon cleanup, or template extraction if those are not already merged.
- Moving code into many tiny modules with unclear value.
- Creating a complex plugin/registry architecture.
- Weakening tests because imports changed.

---

## Files to inspect

Inspect these files before implementing:

- `create_ml_starter.py`
- `templates/**`
- `tests/**`
- `examples/generated-project-sample/**`

---

## Files allowed to change

Only these files or areas are expected to change:

- `create_ml_starter.py`
- new internal generator modules if needed, for example:
  - `ml_starter_generator/__init__.py`
  - `ml_starter_generator/cli.py`
  - `ml_starter_generator/config.py`
  - `ml_starter_generator/scaffold.py`
  - `ml_starter_generator/templates.py`
  - `ml_starter_generator/io.py`
- `tests/**`
- `CHANGELOG.md` only if project convention requires noting the refactor

If another file appears necessary, explain why in the PR description before changing it.

---

## Files not allowed to change

Do not change these files unless the PR explicitly explains why the P0 cannot be completed safely without it:

- root docs unrelated to generator usage;
- ML learning docs under `docs/**`;
- generated project sample content unless needed to prove behavior is unchanged;
- templates unless import/module extraction requires a path adjustment;
- environment requirement versions.

Specific restrictions:

- Do not add dependencies.
- Do not change CLI prompt semantics.
- Do not change generated config schema.
- Do not change generated package layout.
- Do not change task semantics.
- Do not reintroduce Datathon behavior.

---

## Implementation constraints

The implementation must:

- keep the entry command `python create_ml_starter.py` working;
- preserve existing public helper behavior where tests rely on it, or update tests with a clear reason;
- keep modules readable and practical, not over-abstracted;
- avoid global mutable state beyond existing constants unless justified;
- keep template loading standard-library only;
- preserve `force` behavior and output directory safety behavior;
- avoid changing generated output as part of module splitting.

Suggested responsibility split:

```txt
create_ml_starter.py          thin entrypoint
ml_starter_generator/cli.py   prompts and choices
ml_starter_generator/config.py config dict creation
ml_starter_generator/scaffold.py directory/file orchestration
ml_starter_generator/templates.py render/load template helpers
ml_starter_generator/io.py write/touch helpers
```

The final shape can differ if it is simpler and easier to maintain.

---

## Required tests

Add or update targeted tests in the same PR. Tests are part of the deliverable, not optional validation notes.

| P0 | Required proof | Test file / command |
| --- | --- | --- |
| P0.1 | Responsibilities are split without overengineering | code review plus focused tests |
| P0.2 | `python create_ml_starter.py` entrypoint behavior remains available | existing generator tests or entrypoint test |
| P0.3 | Supported generated scenarios still work | `python -m unittest discover -s tests` |
| P0.4 | Import-path changes are reflected in tests without weakening behavior assertions | updated tests |
| P0.5 | No external dependency added | file review plus tests |

Required test cases to preserve:

- [ ] supported task generation;
- [ ] generated config assertions;
- [ ] optional files generation;
- [ ] Python profile generation;
- [ ] Torch requirements generation;
- [ ] output directory safety;
- [ ] package name normalization.

Tests must fail if the generator entrypoint breaks or if supported generated behavior changes.

Do not satisfy this section by only testing private helper functions while public generator behavior remains untested.

---

## Scope and result discipline

Expected changed files:

```txt
create_ml_starter.py, new generator modules, tests/** as needed
```

Expected production change:

```txt
medium internal refactor; generated behavior should stay the same
```

Expected test change:

```txt
update imports/helpers only where needed; keep behavior proof
```

Guidelines:

- focus the implementation on responsibility split;
- preserve existing behavior outside module organization;
- avoid touching files outside allowed areas;
- explain extra touched files in the PR description;
- do not treat formatting-only churn as the deliverable.

A PR is incomplete when generated behavior changes silently, the entrypoint breaks, tests are weakened, or module splitting becomes an unneeded architecture rewrite.

---

## Validation commands

Run:

```bash
python -m py_compile create_ml_starter.py
python -m unittest discover -s tests
python create_ml_starter.py
```

For the interactive command, do not rely on a manual full run if tests already simulate the prompt flow; state how it was validated.

If a command cannot be run, state exactly why in the PR description.

---

## PR description must include

The PR must link this issue using a GitHub closing keyword:

```md
Fixes #20
```

```md
## Summary

[What changed.]

## P0 coverage

- P0.1: [how satisfied]
- P0.2: [how satisfied]
- P0.3: [how satisfied]
- P0.4: [how satisfied]
- P0.5: [how satisfied]

## Test coverage

| P0 / risk | Test proof | Command/result |
| --- | --- | --- |
| Entrypoint preserved | [test/path] | [command/result] |
| Generated behavior preserved | [tests] | [command/result] |
| Compatibility | [existing scenarios] | [command/result] |

## Validation commands

- [command/result]
- [command/result]

## Scope/result notes

Changed files:
- `...`

Extra touched files and why:
- None, or `...`

Weakest proof / known gaps:
- None, or `[specific gap]`

P0 behavior is implemented and tested.
```

---

## Acceptance checklist

- [ ] All P0 objectives satisfied.
- [ ] Generator responsibilities are split clearly.
- [ ] `python create_ml_starter.py` remains the entrypoint.
- [ ] Generated project behavior remains compatible.
- [ ] Tests continue proving supported scenarios.
- [ ] No new dependencies added.
- [ ] No unrelated feature work included.
- [ ] No Datathon behavior reintroduced.
- [ ] Extra touched files are justified by the P0 result.
- [ ] PR links the issue with `Fixes #20` or `Closes #20`.
